### ISSUE TITLE
cmd/go: panic: runtime error: index out of range [-1] in collectDepsErrors

### ISSUE THREAD

**Issue by findleyr** [golang/go#61816]
Related: #59905.

Just observed this panic while looking into golang/vscode-go#2923:

From the root directory of  https://github.com/yaklang/yaklang:
```
> go list -e -compiled -test ./...
panic: runtime error: index out of range [-1]

goroutine 1 [running]:
cmd/go/internal/list.collectDepsErrors.func1(0xc0029afae8?, 0xc001d51008?)
        /home/rfindley/src/go/src/cmd/go/internal/list/list.go:962 +0x134
sort.insertionSort_func({0xc001d51210?, 0xc001bde8a0?}, 0x0, 0x2)
        /home/rfindley/src/go/src/sort/zsortfunc.go:12 +0xa7
sort.pdqsort_func({0xc001d51210?, 0xc001bde8a0?}, 0x7f2b023bc108?, 0x18?, 0xc000506000?)
        /home/rfindley/src/go/src/sort/zsortfunc.go:73 +0x31b
sort.Slice({0x9af040?, 0xc0029afae8?}, 0x2?)
        /home/rfindley/src/go/src/sort/slice.go:29 +0xf8
cmd/go/internal/list.collectDepsErrors(0xc002898600)
        /home/rfindley/src/go/src/cmd/go/internal/list/list.go:951 +0x2f7
cmd/go/internal/list.runList({0xb896b8?, 0xf2cfc0}, 0xef0a00?, {0xc000022230?, 0x1, 0x1})
        /home/rfindley/src/go/src/cmd/go/internal/list/list.go:807 +0x21c5
main.invoke(0xef0a00, {0xc0000221f0, 0x5, 0x5})
        /home/rfindley/src/go/src/cmd/go/main.go:268 +0x5c7
main.main()
        /home/rfindley/src/go/src/cmd/go/main.go:186 +0x786
```

Does not reproduce on 1.20.

@matloob @bcmills should this be a release blocker? It breaks gopls, and I'm not sure how many repositories it affects.

**Comment by findleyr** [comment]
CC @golang/release for awareness, as it is the 11th hour (or 11:59pm, perhaps).

Given that others have not encountered this issue, this may be rare enough not to block the 1.21.0 release. But with that said, I'm concerned about shipping a known gopls breakage with 1.21.

**Comment by findleyr** [comment]
Confirmed that this panic also occurs with `go1.21rc4`.

**Comment by bcmills** [comment]
I think we had an omission in https://go.dev/cl/501978: looks like we missed the case wher `len(stkj) == 0` and `len(stki) != 0`.

**Comment by findleyr** [comment]
@gopherbot please back-port this to 1.21. It is a new Go command crash.

**Comment by matloob** [comment]
I have a fix cl. I wasn't able to reproduce using the yaklang repo, but I reproduced it in a test case.

**Comment by findleyr** [comment]
> I have a fix cl. I wasn't able to reproduce using the yaklang repo, but I reproduced it in a test case.

Interesting. For me, it took nothing more than checking out the repo and running `go list -e -compiled -test ./...` from the root directory, reproducing 100% of the time. I wonder if it is related to GOOS=linux?

**Comment by matloob** [comment]
It could be. I just tried it on linux and was able to repro.

**Comment by geitir** [comment]
Hi is this fixed in release version of go1.21 or do we need to wait for a minor version?

**Comment by cagedmantis** [comment]
@geitir This will be included in the next minor release.

**Comment by dreamerlzl** [comment]
Can confirm that this doesn't happen in the just released go 1.21.1 on my side.
